flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinator.java

```diff
@@ -0,0 +1,230 @@
+	private final Source<?, SplitT, EnumChkT> source;
+	/** The serializer that handles the serde of the SplitEnumerator checkpoints. */
+	private final SimpleVersionedSerializer<EnumChkT> enumCheckpointSerializer;
+	/** The serializer for the SourceSplit of the associated Source. */
+	private final SimpleVersionedSerializer<SplitT> splitSerializer;
+	/** The context containing the states of the coordinator. */
+	private final SourceCoordinatorContext<SplitT> context;
+	/** The split enumerator created from the associated Source. */
+	private SplitEnumerator<SplitT, EnumChkT> enumerator;
+	/** A flag marking whether the coordinator has started. */
+	private boolean started;
+
+	public SourceCoordinator(
+			ExecutorService coordinatorExecutor,
+			Source<?, SplitT, EnumChkT> source,
+			SourceCoordinatorContext<SplitT> context) {
+		this.coordinatorExecutor = coordinatorExecutor;
+		this.source = source;
+		this.enumCheckpointSerializer = source.getEnumeratorCheckpointSerializer();
+		this.splitSerializer = source.getSplitSerializer();
+		this.context = context;
+		this.enumerator = source.createEnumerator(context);
+		this.started = false;
+	}
+
+	@Override
+	public void start() throws Exception {
+		LOG.info("Starting split enumerator.");
+		enumerator.start();
+		started = true;
+	}
+
+	@Override
+	public void close() throws Exception {
+		ensureStarted();
+		LOG.info("Closing SourceCoordinator.");
+		enumerator.close();
+		coordinatorExecutor.shutdown();
+		LOG.info("Source coordinator closed.");
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void handleEventFromOperator(int subtask, OperatorEvent event) throws Exception {
+		ensureStarted();
+		coordinatorExecutor.execute(() -> {
+			LOG.debug("Handling event from operator: {}", event);
+			if (event instanceof SourceEventWrapper) {
+				enumerator.handleSourceEvent(subtask, ((SourceEventWrapper) event).getSourceEvent());
+			} else if (event instanceof ReaderRegistrationEvent) {
+				handleReaderRegistrationEvent((ReaderRegistrationEvent) event);
+			}
+		});
+	}
+
+	@Override
+	public void subtaskFailed(int subtaskId) {
+		ensureStarted();
+		coordinatorExecutor.execute(() -> {
+			LOG.info("Handling subtask {} failure.", subtaskId);
+			List<SplitT> splitsToAddBack = context.getAndRemoveUncheckpointedAssignment(subtaskId);
+			context.unregisterSourceReader(subtaskId);
+			LOG.debug("Adding {} back to the split enumerator.", splitsToAddBack);
+			enumerator.addSplitsBack(splitsToAddBack, subtaskId);
+		});
+	}
+
+	@Override
+	public CompletableFuture<byte[]> checkpointCoordinator(long checkpointId) throws Exception {
+		ensureStarted();
+		return CompletableFuture.supplyAsync(() -> {
+			try {
+				LOG.debug("Taking a state snapshot for checkpoint {}", checkpointId);
+				return toBytes(checkpointId);
+			} catch (Exception e) {
+				throw new FlinkRuntimeException("Failed to checkpoint coordinator due to ", e);
```

Comment:
It would be nice to avoid exception wrapping when possible. To make checked exceptions work with Futures, I think the way to go is to wrap it in a `CompletionException` which gets automatically unwrapped when accessing the original exception.